executor/merge_join.go

```diff
@@ -43,169 +41,225 @@ type MergeJoinExec struct {
+		}
+
+		isFirstGroupSameAsPrev, err := t.groupChecker.splitIntoGroups(t.childChunk)
+		if err != nil {
+			return err
+		}
+		if isFirstGroupSameAsPrev && !isEmpty {
+			t.selectNextGroup()
+		}
+	}
+	if isEmpty {
+		goto fetchNext
 	}
 
-	// no more data.
-	if t.firstRow4Key == t.curIter.End() {
-		// the iterator is a blank iterator only as a place holder
-		return chunk.NewIterator4Chunk(t.curChk), nil
+	// iterate all data in t.rowContainer and t.childChunk
+	var iter chunk.Iterator
+	if t.rowContainer.NumChunks() != 0 {
+		iter = chunk.NewIterator4RowContainer(t.rowContainer)
 	}
-	t.selectRow(t.firstRow4Key)
-	for {
-		selectedRow, err := t.nextRow()
-		// error happens or no more data.
-		if err != nil || selectedRow == t.curIter.End() {
-			t.firstRow4Key = t.curIter.End()
-			return t.selectedRowsIter(), err
-		}
-		compareResult := compareChunkRow(t.keyCmpFuncs, selectedRow, t.firstRow4Key, t.joinKeys, t.joinKeys)
-		if compareResult == 0 {
-			t.selectRow(selectedRow)
+	if len(t.groupRowsSelected) != 0 {
+		if iter != nil {
+			iter = chunk.NewMultiIterator(iter, t.childChunkIter)
 		} else {
-			t.firstRow4Key = selectedRow
-			return t.selectedRowsIter(), nil
+			iter = t.childChunkIter
 		}
 	}
+	t.groupRowsIter = iter
+	t.groupRowsIter.Begin()
+	return nil
 }
 
-func (t *mergeJoinInnerTable) nextRow() (chunk.Row, error) {
-	for {
-		if t.curRow == t.curIter.End() {
-			err := t.reallocCurChkForReader()
-			if err != nil {
-				t.curRow = t.curIter.End()
-				return t.curRow, err
-			}
-			oldMemUsage := t.curChk.MemoryUsage()
-			err = Next(t.ctx, t.reader, t.curChk)
-			// error happens or no more data.
-			if err != nil || t.curChk.NumRows() == 0 {
-				t.curRow = t.curIter.End()
-				return t.curRow, err
-			}
-			newMemUsage := t.curChk.MemoryUsage()
-			t.memTracker.Consume(newMemUsage - oldMemUsage)
-			t.curRow = t.curIter.Begin()
+func (t *mergeJoinTable) fetchNextOuterGroup(ctx context.Context, exec *MergeJoinExec, requiredRows int) error {
+	if t.executed && t.groupChecker.isExhausted() {
+		return nil
+	}
+
+	if !t.executed && t.groupChecker.isExhausted() {
+		// It's hard to calculate selectivity if there is any filter or it's inner join,
+		// so we just push the requiredRows down when it's outer join and has no filter.
+		if exec.isOuterJoin && len(t.filters) == 0 {
+			t.childChunk.SetRequiredRows(requiredRows, exec.maxChunkSize)
+		}
+		err := t.fetchNextChunk(ctx, exec)
+		if err != nil || t.executed {
+			return err
 		}
 
-		result := t.curRow
-		t.curChkInUse = true
-		t.curRow = t.curIter.Next()
+		t.childChunkIter.Begin()
+		t.filtersSelected, err = expression.VectorizedFilter(exec.ctx, t.filters, t.childChunkIter, t.filtersSelected)
+		if err != nil {
+			return err
+		}
 
-		if !t.hasNullInJoinKey(result) {
-			return result, nil
+		_, err = t.groupChecker.splitIntoGroups(t.childChunk)
+		if err != nil {
+			return err
 		}
 	}
+
+	t.groupRowsSelected = t.groupRowsSelected[:0]
+	t.selectNextGroup()
```

Comment:
How about initializing the `t.groupRowsSelected` in the method `selectNextGroup` so that we don't need to initialize it manually each time?

Reply:
Good idea.

Reply:
fixed